recipes/LibriSpeech/ASR/CTC/train_with_wav2vec_k2.py

```diff
@@ -0,0 +1,557 @@
+
+    # test is separate
+    test_datasets = {}
+    for csv_file in hparams["test_csv"]:
+        name = Path(csv_file).stem
+        test_datasets[name] = sb.dataio.dataset.DynamicItemDataset.from_csv(
+            csv_path=csv_file, replacements={"data_root": data_folder}
+        )
+        test_datasets[name] = test_datasets[name].filtered_sorted(
+            sort_key="duration"
+        )
+
+    datasets = [train_data, valid_data] + [i for k, i in test_datasets.items()]
+
+    # 2. Define audio pipeline:
+    @sb.utils.data_pipeline.takes("wav")
+    @sb.utils.data_pipeline.provides("sig")
+    def audio_pipeline(wav):
+        sig = sb.dataio.dataio.read_audio(wav)
+        return sig
+
+    sb.dataio.dataset.add_dynamic_item(datasets, audio_pipeline)
+
+    # 3. Define text pipeline:
+    @sb.utils.data_pipeline.takes("wrd")
+    @sb.utils.data_pipeline.provides("wrd", "char_list")
+    def text_pipeline(wrd):
+        yield wrd
+        char_list = list(wrd)
+        yield char_list
+
+    sb.dataio.dataset.add_dynamic_item(datasets, text_pipeline)
+
+    # 4. Set output:
+    sb.dataio.dataset.set_output_keys(
+        datasets, ["id", "sig", "wrd", "char_list"],
+    )
+
+    return train_data, valid_data, test_datasets
+
+
+if __name__ == "__main__":
+    # CLI:
+    hparams_file, run_opts, overrides = sb.parse_arguments(sys.argv[1:])
+
+    # If distributed_launch=True then
+    # create ddp_group with the right communication protocol
+    sb.utils.distributed.ddp_init_group(run_opts)
+
+    with open(hparams_file) as fin:
+        hparams = load_hyperpyyaml(fin, overrides)
+
+    # env_corrupt is not supported with k2 yet
+    if hparams.get("env_corrupt", None):
+        raise NotImplementedError("env_corrupt is not supported with k2 yet")
+
+    # Create experiment directory
+    sb.create_experiment_directory(
+        experiment_directory=hparams["output_folder"],
+        hyperparams_to_save=hparams_file,
+        overrides=overrides,
+    )
+
+    # Dataset prep (parsing Librispeech)
+    import librispeech_prepare
+
+    # multi-gpu (ddp) save data preparation
+    run_on_main(
+        librispeech_prepare.prepare_librispeech,
+        kwargs={
+            "data_folder": hparams["data_folder"],
+            "tr_splits": hparams["train_splits"],
+            "dev_splits": hparams["dev_splits"],
+            "te_splits": hparams["test_splits"],
+            "save_folder": hparams["output_folder"],
+            "merge_lst": hparams["train_splits"],
+            "merge_name": "train.csv",
+            "skip_prep": hparams["skip_prep"],
+        },
+    )
+
+    run_on_main(
+        librispeech_prepare.download_librispeech_vocab_text,
+        kwargs={"destination": hparams["data_folder"]},
+    )
+
+    # here we create the datasets objects as well as tokenization and encoding
+    train_data, valid_data, test_datasets = dataio_prepare(hparams)
+
+    # Create the lexicon.txt for k2
+    run_on_main(
```

Comment:
@Adel-Moumen the recipe should be tested with DDP. Lots of file prep, so we must make sure that it works.

Reply:
OK. WIll do.

Reply:
Hello @TParcollet, could you please let me know how to test our recipe with DDP, or possibly give me a recipe which has been tested with DDP for my reference? I tried to run our recipe with DDP but have not been successful yet. 

```
$ python -m torch.distributed.launch --nproc_per_node=2 train_with_wav2vec_k2.py  hparams/train_hf_wav2vec_k2.yaml --data_folder $LOCAL_HOME/data/librispeech/LibriSpeech --seed 1234 --compose_HL_with_G false --distributed_launch
/disk/scratch3/zzhao/conda/sb/lib/python3.8/site-packages/torch/distributed/launch.py:181: FutureWarning: The module torch.distributed.launch is deprecated
and will be removed in future. Use torchrun.
Note that --use-env is set by default in torchrun.
If your script expects `--local-rank` argument to be set, please
change it to read from `os.environ['LOCAL_RANK']` instead. See
https://pytorch.org/docs/stable/distributed.html#launch-utility for
further instructions

  >warnings.warn(
>WARNING:torch.distributed.run:
*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed.
*****************************************
torchvision is not available - cannot save figures
torchvision is not available - cannot save figures
Traceback (most recent call last):
  File "train_with_wav2vec_k2.py", line 380, in <module>
    hparams = load_hyperpyyaml(fin, overrides)
  File "/disk/scratch3/zzhao/conda/sb/lib/python3.8/site-packages/hyperpyyaml/core.py", line 157, in load_hyperpyyaml
    yaml_stream = resolve_references(yaml_stream, overrides, overrides_must_match)
  File "/disk/scratch3/zzhao/conda/sb/lib/python3.8/site-packages/hyperpyyaml/core.py", line 316, in resolve_references
    recursive_update(preview, overrides, must_match=overrides_must_match)
  File "/disk/scratch3/zzhao/conda/sb/lib/python3.8/site-packages/hyperpyyaml/core.py", line 776, in recursive_update
    raise KeyError(f"Override '{k}' not found in: {[key for key in d.keys()]}")
KeyError: "Override 'local-rank' not found in: ['seed', '__set_seed', 'output_folder', 'save_folder', 'train_log', 'wav2vec2_hub', 'wav2vec2_folder', 'data_folder', 'train_splits', 'dev_splits', 'test_splits', 'skip_prep', 'ckpt_interval_minutes', 'train_csv', 'valid_csv', 'test_csv', 'lang_dir', 'vocab_file', 'sil_prob', 'add_word_boundary', 'test_search_beam', 'test_output_beam', 'test_min_active_state', 'test_max_active_state', 'ac_scale', 'compose_HL_with_G', 'decoding_method', 'rescoring_lm_scale', 'lm_dir', 'G_arpa', 'G_rescoring_arpa', 'number_of_epochs', 'lr', 'lr_wav2vec', 'sorting', 'precision', 'sample_rate', 'batch_size', 'test_batch_size', 'num_workers', 'train_dataloader_opts', 'valid_dataloader_opts', 'test_dataloader_opts', 'activation', 'dnn_layers', 'dnn_neurons', 'freeze_wav2vec', 'output_neurons', 'epoch_counter', 'augmentation', 'enc', 'wav2vec2', 'ctc_lin', 'log_softmax', 'ctc_cost', 'modules', 'model', 'model_opt_class', 'wav2vec_opt_class', 'lr_annealing_model', 'lr_annealing_wav2vec', 'checkpointer', 'train_logger', 'error_rate_computer', 'cer_computer']"
Traceback (most recent call last):
  File "train_with_wav2vec_k2.py", line 380, in <module>
    hparams = load_hyperpyyaml(fin, overrides)
  File "/disk/scratch3/zzhao/conda/sb/lib/python3.8/site-packages/hyperpyyaml/core.py", line 157, in load_hyperpyyaml
    yaml_stream = resolve_references(yaml_stream, overrides, overrides_must_match)
  File "/disk/scratch3/zzhao/conda/sb/lib/python3.8/site-packages/hyperpyyaml/core.py", line 316, in resolve_references
    recursive_update(preview, overrides, must_match=overrides_must_match)
  File "/disk/scratch3/zzhao/conda/sb/lib/python3.8/site-packages/hyperpyyaml/core.py", line 776, in recursive_update
    raise KeyError(f"Override '{k}' not found in: {[key for key in d.keys()]}")
KeyError: "Override 'local-rank' not found in: ['seed', '__set_seed', 'output_folder', 'save_folder', 'train_log', 'wav2vec2_hub', 'wav2vec2_folder', 'data_folder', 'train_splits', 'dev_splits', 'test_splits', 'skip_prep', 'ckpt_interval_minutes', 'train_csv', 'valid_csv', 'test_csv', 'lang_dir', 'vocab_file', 'sil_prob', 'add_word_boundary', 'test_search_beam', 'test_output_beam', 'test_min_active_state', 'test_max_active_state', 'ac_scale', 'compose_HL_with_G', 'decoding_method', 'rescoring_lm_scale', 'lm_dir', 'G_arpa', 'G_rescoring_arpa', 'number_of_epochs', 'lr', 'lr_wav2vec', 'sorting', 'precision', 'sample_rate', 'batch_size', 'test_batch_size', 'num_workers', 'train_dataloader_opts', 'valid_dataloader_opts', 'test_dataloader_opts', 'activation', 'dnn_layers', 'dnn_neurons', 'freeze_wav2vec', 'output_neurons', 'epoch_counter', 'augmentation', 'enc', 'wav2vec2', 'ctc_lin', 'log_softmax', 'ctc_cost', 'modules', 'model', 'model_opt_class', 'wav2vec_opt_class', 'lr_annealing_model', 'lr_annealing_wav2vec', 'checkpointer', 'train_logger', 'error_rate_computer', 'cer_computer']"
ERROR:torch.distributed.elastic.multiprocessing.api:failed (exitcode: 1) local_rank: 0 (pid: 210993) of binary: /disk/scratch3/zzhao/conda/sb/bin/python
Traceback (most recent call last):
  File "/disk/scratch3/zzhao/conda/sb/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/disk/scratch3/zzhao/conda/sb/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/disk/scratch3/zzhao/conda/sb/lib/python3.8/site-packages/torch/distributed/launch.py", line 196, in <module>
    main()
  File "/disk/scratch3/zzhao/conda/sb/lib/python3.8/site-packages/torch/distributed/launch.py", line 192, in main
    launch(args)
  File "/disk/scratch3/zzhao/conda/sb/lib/python3.8/site-packages/torch/distributed/launch.py", line 177, in launch
    run(args)
  File "/disk/scratch3/zzhao/conda/sb/lib/python3.8/site-packages/torch/distributed/run.py", line 785, in run
    elastic_launch(
  File "/disk/scratch3/zzhao/conda/sb/lib/python3.8/site-packages/torch/distributed/launcher/api.py", line 134, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/disk/scratch3/zzhao/conda/sb/lib/python3.8/site-packages/torch/distributed/launcher/api.py", line 250, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError:
============================================================
train_with_wav2vec_k2.py FAILED
------------------------------------------------------------

```

Reply:
Hi @ZhaoZeyu1995, you need to run the following command : `torch run  --nproc_per_node=2 train_with_wav2vec_k2.py  hparams/train_hf_wav2vec_k2.yaml --data_folder $LOCAL_HOME/data/librispeech/LibriSpeech --seed 1234 --compose_HL_with_G false`.